django_irods/s3_backend.py

```diff
@@ -0,0 +1,714 @@
+    from botocore.exceptions import ClientError
+    from botocore.signers import CloudFrontSigner
+except ImportError as e:
+    raise ImproperlyConfigured("Could not load Boto3's S3 bindings. %s" % e)
+
+
+# NOTE: these are defined as functions so both can be tested
+def _use_cryptography_signer():
+    # https://cryptography.io as an RSA backend
+    from cryptography.hazmat.backends import default_backend
+    from cryptography.hazmat.primitives import hashes
+    from cryptography.hazmat.primitives.asymmetric import padding
+    from cryptography.hazmat.primitives.serialization import load_pem_private_key
+
+    def _cloud_front_signer_from_pem(key_id, pem):
+        if isinstance(pem, str):
+            pem = pem.encode("ascii")
+        key = load_pem_private_key(pem, password=None, backend=default_backend())
+
+        return CloudFrontSigner(
+            key_id, lambda x: key.sign(x, padding.PKCS1v15(), hashes.SHA1())
+        )
+
+    return _cloud_front_signer_from_pem
+
+
+def _use_rsa_signer():
+    # https://stuvel.eu/rsa as an RSA backend
+    import rsa
+
+    def _cloud_front_signer_from_pem(key_id, pem):
+        if isinstance(pem, str):
+            pem = pem.encode("ascii")
+        key = rsa.PrivateKey.load_pkcs1(pem)
+        return CloudFrontSigner(key_id, lambda x: rsa.sign(x, key, "SHA-1"))
+
+    return _cloud_front_signer_from_pem
+
+
+for _signer_factory in (_use_cryptography_signer, _use_rsa_signer):
+    try:
+        _cloud_front_signer_from_pem = _signer_factory()
+        break
+    except ImportError:
+        pass
+else:
+
+    def _cloud_front_signer_from_pem(key_id, pem):
+        raise ImproperlyConfigured(
+            "An RSA backend is required for signing cloudfront URLs.\n"
+            "Supported backends are packages: cryptography and rsa."
+        )
+
+
+def _filter_download_params(params):
+    return {
+        key: value
+        for (key, value) in params.items()
+        if key in s3transfer.constants.ALLOWED_DOWNLOAD_ARGS
+    }
+
+
+@deconstructible
+class S3File(CompressedFileMixin, File):
+    """
+    The default file object used by the S3Storage backend.
+
+    This file implements file streaming using boto's multipart
+    uploading functionality. The file can be opened in read or
+    write mode.
+
+    This class extends Django's File class. However, the contained
+    data is only the data contained in the current buffer. So you
+    should not access the contained file object directly. You should
+    access the data via this class.
+
+    Warning: This file *must* be closed using the close() method in
+    order to properly write the file to S3. Be sure to close the file
+    in your application.
+    """
+
+    def __init__(self, name, mode, storage, buffer_size=None):
+        if "r" in mode and "w" in mode:
+            raise ValueError("Can't combine 'r' and 'w' in mode.")
+        self._storage = storage
+        bucket, name = bucket_and_name(name)
```

Comment:
This whole file was ripped from django-storages s3 implementation with some involved updates to support multiple buckets

Reply:
Just a thought: could we fork `django-storages` into hydroshare org and make the updates on that fork as monkeypatch, so that it is easier to pull in there updates as they make them?

Reply:
I think we could probably just get away with extending the S3File and S3Storage classes from the django-storages module. Basically just update this file to extend those classes here and only declare the changed methods.

Reply:
The last 2 commits make the proposed change.

Reply:
NICE!

Reply:
Tests aren't passing with it. I reverted those commits and will work on it in a branch that will be merged to here once it is worked out.

Reply:
I figured it out and it is now included with this PR